featurebyte/query_graph/sql/feature_historical.py

```diff
@@ -365,33 +366,68 @@ def get_historical_features_expr(
 
 
 def split_nodes(
+    graph: QueryGraph,
     nodes: list[Node],
     num_features_per_query: int,
+    is_tile_cache: bool = False,
 ) -> list[list[Node]]:
     """
     Split nodes into multiple lists, each containing at most `num_features_per_query` nodes. Nodes
     within the same group after splitting will be executed in the same query.
 
     Parameters
     ----------
+    graph: QueryGraph
+        Query graph
     nodes : list[Node]
         List of nodes
     num_features_per_query : int
         Number of features per query
+    is_tile_cache : bool
+        Whether the output will be used for tile cache queries
 
     Returns
     -------
     list[list[Node]]
     """
+    planner = FeatureExecutionPlanner(graph=graph, is_online_serving=False)
+
+    def get_sort_key(node: Node) -> str:
+        mapped_node = planner.graph.get_node_by_name(planner.node_name_map[node.name])
+        agg_specs = planner.get_aggregation_specs(mapped_node)
+        agg_spec = agg_specs[0]
+
+        parts = [agg_spec.aggregation_type.value]
+        if isinstance(agg_spec, TileBasedAggregationSpec):
+            if is_tile_cache:
+                # Tile cache queries joins with entity tracker tables. These tables are organized by
+                # aggregation_id. The split should be random across different aggregation_id.
+                parts.append(agg_spec.aggregation_id)
+            else:
+                # Tile based aggregation joins with tile tables. Sort by tile_table_id first to
+                # group nodes that join with the same tile table.
+                parts.extend([agg_spec.tile_table_id, agg_spec.aggregation_id])
+        else:
+            assert isinstance(agg_spec, NonTileBasedAggregationSpec)
+            # These queries join with source tables directly. Sort by query node name of the source
+            # to group nodes that join with the same source table.
+            query_node = planner.graph.get_node_by_name(agg_spec.aggregation_source.query_node_name)
+            parts.append(query_node.name)
+
+        key = ",".join(parts)
+        return key
+
     result = []
-    for i in range(0, len(nodes), num_features_per_query):
-        current_nodes = nodes[i : i + num_features_per_query]
+    sorted_nodes = sorted(nodes, key=get_sort_key)
+    for i in range(0, len(sorted_nodes), num_features_per_query):
+        current_nodes = sorted_nodes[i : i + num_features_per_query]
         result.append(current_nodes)
```

Comment:
Would you expect to see performance improvement if we could ensure that same source table should be within the same node group (if the size of the group is less than `num_features_per_query`)?

Reply:
If the size of the group is less than `num_features_per_query`, all the nodes would have been in the same group already regardless of source table, so I think I'm misunderstanding the question. Could you clarify a bit more?